OpenRA.Mods.Common/Orders/RepairOrderGenerator.cs

```diff
@@ -12,86 +12,191 @@
 			{
-				VisualFeedbackTarget = Target.FromActor(underCursor)
+				VisualFeedbackTarget = Target.FromActor(actor)
 			};
 		}
 
-		protected override void Tick(World world)
+		protected override IEnumerable<IRenderable> Render(WorldRenderer wr, World world) { yield break; }
+		protected override IEnumerable<IRenderable> RenderAboveShroud(WorldRenderer wr, World world) { yield break; }
+		protected override IEnumerable<IRenderable> RenderAnnotations(WorldRenderer wr, World world)
 		{
-			if (world.LocalPlayer != null &&
-				world.LocalPlayer.WinState != WinState.Undefined)
-				world.CancelInputMode();
+			if (isDragging && (mousePos - dragStartMousePos).Length > Game.Settings.Game.SelectionDeadzone)
+			{
+				var lastMousePos = wr.Viewport.ViewToWorldPx(Viewport.LastMousePos);
+				var diag1 = wr.ProjectedPosition(lastMousePos);
+				var diag2 = wr.ProjectedPosition(dragStartMousePos);
+				yield return new RectangleAnnotationRenderable(diag1, diag2, diag1, 1, Color.White);
+				var actors = SelectActorsInBoxWithDeadzone(world, dragStartMousePos, lastMousePos);
+
+				if (actors != null && actors.Any())
+					foreach (var actor in actors)
+					{
+						if (actor != null)
+						{
+							var decorationBounds = actor.TraitsImplementing<IDecorationBounds>().ToArray();
+							var bounds = decorationBounds.FirstNonEmptyBounds(actor, wr);
+							yield return new SelectionBarsAnnotationRenderable(actor, bounds, true, false);
+						}
+					}
+			}
+
+			yield break;
 		}
 
-		protected override IEnumerable<IRenderable> Render(WorldRenderer wr, World world) { yield break; }
-		protected override IEnumerable<IRenderable> RenderAboveShroud(WorldRenderer wr, World world) { yield break; }
-		protected override IEnumerable<IRenderable> RenderAnnotations(WorldRenderer wr, World world) { yield break; }
+		static IEnumerable<Actor> SelectActorsInBoxWithDeadzone(World world, int2 a, int2 b)
+		{
+			// For dragboxes that are too small, choose nothing
+			// because the WorldInteractionControllerWidget can show the imformation
+			if ((a - b).Length <= Game.Settings.Game.SelectionDeadzone)
+				return null;
+
+			return world.ScreenMap.ActorsInMouseBox(a, b)
+					.Select(x => x.Actor)
+					.Where(x => x.AppearsFriendlyTo(world.LocalPlayer.PlayerActor) && !world.FogObscures(x));
+		}
 
 		protected override string GetCursor(World world, CPos cell, int2 worldPixel, MouseInput mi)
 		{
-			mi.Button = MouseButton.Left;
-			return OrderInner(world, mi).Any()
-				? "repair" : "repair-blocked";
+			// DnqBob: Only when mouse over unrepairable actor, the cursor will change to "repair-blocked"
+			if (MouseOverActor(world, worldPixel, mi))
+				return "repair";
+			else
+				return "repair-blocked";
+		}
+
+		static bool MouseOverActor(World world, int2 worldPixel, MouseInput mi)
+		{
+			var actors = world.ScreenMap.ActorsAtMouse(worldPixel)
+					.Select(x => x.Actor)
+					.Where(x => !world.FogObscures(x));
+			if (actors == null || !actors.Any())
```

Comment:
Style nit: This can be simplified to:
```
var actors = world.ScreenMap.ActorsAtMouse(worldPixel)
			.Select(x => x.Actor)
			.Where(x => !world.FogObscures(x));

return !actors.Any() || RepairUnderCondition(actors.First(), world, mi) != null;
```

Reply:
OK